Add agent idv0.9.0

cyberdrk opened the pull request:
added a functionality to acquire agentid and default arg to AlicaContext, AlicaEngine and TeamManager Constructors 

If the agentid is null, only then does the TeamManager read it from Local.conf file or generate it. 
If the agentid is present (passed to AlicaContext), then the TeamManager does not bother doing the above.

Commit:
added a functionality to acquire agentid and default arg to constructor declarations

Commit:
added changes after testing todo: code cleanup

Commit:
Removed ROS_INFO statements and ros.h

cyberdrk commented:
todo: add unit test for the same

veerajsk commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -125,7 +126,7 @@ class AlicaContext
      * @note This is the main alica api class
      */
-    AlicaContext(const std::string& roleSetName, const std::string& masterPlanName, bool stepEngine);
+    AlicaContext(const std::string& roleSetName, const std::string& masterPlanName, bool stepEngine, AgentIDConstPtr agentID = nullptr);
```
Would be better to take the agent Id as an integer parameter instead of agentId class, considering that the agent id in the alica config file is also specified to be an integer. It would also be easier for the user to use the API.

StephanOpfer replied:
In the case of a long or integer id, the default value cannot be nullptr. As a consequence, we would need to forbid some specific id value (e.g. -1) to be chosen by the application, in order to signal to the Alica engine that no id was set, if the value is -1. 

Therefore, I think that a null-able type is safer. Although the application needs to create an AgentIDConstPtr, which isn't that easy.

abhi1287 replied:
Well, uint64 could be a valid type and 0 meaning nothing was given. But didn't we thought of taking AgentId object instead which internally could be empty? So application can pass something like AlicaContext(......., AgentId(3345) ) .
We might have to overload AgentId constructor to do this. And as far as I remember the bool operator is already overloaded.

StephanOpfer replied:
Added changes as discussed and documented in https://docs.google.com/presentation/d/13TxKVMJgYloT6HifuS5WFgAE8iFoJ88KBbJBlU8HOrQ/edit#slide=id.g72a0f35744_0_6

StephanOpfer reviewed (changes requested):
I tried to compile this PR with the other alica repos. It only compiles with the rr-devel branch of alica-essentials and does not compile with the v0.9 branch of alica-essentials.

The difference is a fix for storing long values in the system_config done on the rr-devel branch, but not on the v0.9 branch. Either we add another PR from rr-devel to v0.9 in alica-essentials or we need to adapt this PR here.

StephanOpfer commented:
Okay I created a PR for alica-essentials with the same name as this one. Together they are compiling.

Commit:
changed constructor's agent parameter to AgentID* and copied the given AgentID* via the _agentIDManager

abhi1287 commented on alica_engine/src/engine/teammanager/TeamManager.cpp:
```diff
@@ -53,12 +53,13 @@ bool AgentsCache::addAgent(Agent* agent)
         , _agentAnnouncementTimeInterval(AlicaTime::zero())
         , _timeLastAnnouncement(AlicaTime::zero())
         , _announcementRetries(0)
+        , _agentId(agentID)
```
don't need to store it, the member Agent* _localAgent will already have the id once constructor finishes execution.
we can pass it to readselffromconfig() as an argument

abhi1287 commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -125,7 +126,7 @@ class AlicaContext
      * @note This is the main alica api class
      */
-    AlicaContext(const std::string& roleSetName, const std::string& masterPlanName, bool stepEngine);
+    AlicaContext(const std::string& roleSetName, const std::string& masterPlanName, bool stepEngine, AgentID* agentID = nullptr);
```
const AgentID*

Commit:
- made agentID a const AgentID*
- fixed segfault in the list of initialisers of the AlicaEngine constructor (horrible line for TeamManager gets removed, when IDManager is owned by AlicaContext in future)

veerajsk commented on alica_engine/src/engine/teammanager/TeamManager.cpp:
```diff
@@ -82,28 +82,32 @@ void TeamManager::setTeamTimeout(AlicaTime t)
             }
         }
+    } else {
+        _localAnnouncement.senderID = agentID;
```
The engine's getId() & _generateId() methods both internally store the ID's in the AgentIDManager but neither is called in the else condition. The AgentIDManager currently does not seem to use the stored ID's but it is better to maintain consistent behaviour.

StephanOpfer commented:
Its stored in the constructor of the Alica Engine. That will be changed in
future, when the AgentIDManager might be inside the AlicaContext.

On Fri, 3 Apr 2020, 19:52 veerajsk <notifications@github.com> wrote:

> *@veerajsk* commented on this pull request.
> ------------------------------
>
> In alica_engine/src/engine/teammanager/TeamManager.cpp
> <https://github.com/rapyuta-robotics/alica/pull/84#discussion_r403193700>:
>
> >              }
>          }
> +    } else {
> +        _localAnnouncement.senderID = agentID;
>
> The engine's getId() & _generateId() methods both internally store the
> ID's in the AgentIDManager but neither is called in the else condition. The
> AgentIDManager currently does not seem to use the stored ID's but it is
> better to maintain consistent behaviour.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/rapyuta-robotics/alica/pull/84#pullrequestreview-387464056>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAW6FOVIM6YOZJIDH7IQKOTRKYO7TANCNFSM4KY4LSPQ>
> .
>

veerajsk commented:
@StephanOpfer yes my mistake I misunderstood the context. I have approved the changes
